### PR TITLE
Adding routing for notifications, a fix for allowing the notifications popup all of the time.

### PR DESCRIPTION
**Pull request by jpw547-zz**

### CODE REVIEW

File: fakebook/src/app/component/like-view/like-view.component.ts [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
   templateUrl: './like-view.component.html',
   styleUrls: ['./like-view.component.css'],
 })
 export class LikeViewComponent {
-  @Input() count!: number;
-  @Input() postId!: number;
-  @Input() liked!: boolean;
+  @Input() count: number | undefined;
+  @Input() postId: number;
+  @Input() liked: boolean;
```
Comment (escalonn) [inline fakebook/src/app/component/like-view/like-view.component.ts]: add a comment explaining when/why the input is expected to be undefined and when/why a number.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts [hunk 2]
```diff
@@ -12,22 +13,43 @@ import { PostService } from '../../service/post.service';
...
   ngOnInit(): void {
     this.getUser();
-    this.getPosts();
+    //this.getPosts();
   }
 
   getPosts(): void {
     this.newsfeedService
-      .getPosts()
+      .getPosts(this.postId)
```
Comment (escalonn) [inline fakebook/src/app/component/newsfeed/newsfeed.component.ts]: why are you passing a postId that's probably undefined here?

File: fakebook/tsconfig.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
     "outDir": "./dist/out-tsc",
     "forceConsistentCasingInFileNames": true,
     "strict": true,
+    "strictPropertyInitialization": false,
```
Comment (escalonn) [inline fakebook/tsconfig.json]: remove this and continue to write code that passes all the strict type checks. i can assist with any confusing cases.

File: fakebook/src/app/services/newsfeed.service.ts [hunk 4]
```diff
@@ -17,7 +17,7 @@ export class NewsfeedService {
     Accept: 'application/json',
   };
 
-  getPosts(): Observable<Post[]> {
+  getPosts(postId?: number): Observable<Post[]> {
```
Comment (escalonn) [inline fakebook/src/app/services/newsfeed.service.ts]: why a useless parameter?

File: fakebook/src/app/component/notifications-view/notifications-view.component.html [hunk 5]
```diff
@@ -47,24 +47,40 @@
...
+        </div>
+      </a>
     </div>
     <div *ngIf="notif.type == 'follow'" style="padding-left: 12px">
-      Here is person {{ notif.userId }} following you<br />{{
-        notif.date.toString() | timeAgo
-      }}
+      <a [routerLink]="['/profile', notif.userId]">
+        <div>
+          Here is person {{ notif.userId }} following you<br />{{
```
Comment (escalonn) [inline fakebook/src/app/component/notifications-view/notifications-view.component.html]: let's fix the awkward english in this template while we're here

File: fakebook/src/app/component/notifications-view/notifications-view.component.css [hunk 6]
```diff
@@ -0,0 +1,8 @@
+a {
+    color: inherit;
+    text-decoration: inherit;
+}
+
+.notification:hover {
+    background-color: lightgrey;
+}
```
Comment (escalonn) [inline fakebook/src/app/component/notifications-view/notifications-view.component.css]: conform to editorconfig: 2-space indent, newline at end of file.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts [hunk 7]
```diff
@@ -12,22 +13,43 @@ import { PostService } from '../../service/post.service';
...
+      .subscribe((gotPosts) => (this.posts = gotPosts));
+  }
+
+  getPostsById(): void {
```
Comment (escalonn) [inline fakebook/src/app/component/newsfeed/newsfeed.component.ts]: why a plural method name when it's going to be just one post?

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts [hunk 8]
```diff
@@ -12,22 +13,43 @@ import { PostService } from '../../service/post.service';
...
+      else{
+        this.getPosts();
+      }
+    })
+  }
 
   ngOnInit(): void {
     this.getUser();
-    this.getPosts();
+    //this.getPosts();
```
Comment (escalonn) [inline fakebook/src/app/component/newsfeed/newsfeed.component.ts]: why is this commented out?

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts [hunk 9]
```diff
@@ -12,22 +13,43 @@ import { PostService } from '../../service/post.service';
...
+    this.route.params.subscribe(params => {
+      console.log(params);
+      if(params['id'] != undefined){   
+        this.postId = +params['id'];
+          if(this.postId != undefined)
+            this.getPostsById();
+      }
+      else{
+        this.getPosts();
+      }
```
Comment (escalonn) [inline fakebook/src/app/component/newsfeed/newsfeed.component.ts]: fix formatting, use strict equals, and consider what will happen if someone types in "newsfeed/asdf": postId will become NaN and we'll send NaN to the server.

File: fakebook/src/app/services/newsfeed.service.ts [hunk 10]
```diff
@@ -29,6 +29,22 @@ export class NewsfeedService {
...
+    };
+    let post: Post;
+    return this.http.get<Post>(
+      `${environment.baseUrls.posts}/api/posts/${postId}`,
+      { headers }
+    ).toPromise().then(p => {
+      post = p;    
+      return [post];
+    });
+  }
```
Comment (escalonn) [inline fakebook/src/app/services/newsfeed.service.ts:42]: why are you converting it to a promise and wrapping it in an array? that's silly. if you want one post it should be an observable of one post.

**Review by escalonn — CHANGES_REQUESTED** [review]
don't request review when the build fails, let alone when the code doesn't even compile. that tells me you're not even running ng serve to check if what you typed is good.

File: fakebook/tsconfig.json (hunk 3 above)
Comment (jpw547-zz) [reply fakebook/tsconfig.json]: Removed.

File: fakebook/src/app/component/like-view/like-view.component.ts (hunk 1 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/like-view/like-view.component.ts]: Fixed.

File: fakebook/src/app/services/newsfeed.service.ts (hunk 4 above)
Comment (jpw547-zz) [reply fakebook/src/app/services/newsfeed.service.ts]: Removed.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts (hunk 2 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/newsfeed/newsfeed.component.ts]: Removed parameter.

File: fakebook/src/app/component/notifications-view/notifications-view.component.html (hunk 5 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/notifications-view/notifications-view.component.html]: Changed.

File: fakebook/src/app/component/notifications-view/notifications-view.component.css (hunk 6 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/notifications-view/notifications-view.component.css]: Fixed.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts (hunk 7 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/newsfeed/newsfeed.component.ts]: Renamed method.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts (hunk 8 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/newsfeed/newsfeed.component.ts]: Deleted this line.

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts (hunk 9 above)
Comment (jpw547-zz) [reply fakebook/src/app/component/newsfeed/newsfeed.component.ts]: Fixed.

File: fakebook/src/app/services/newsfeed.service.ts (hunk 10 above)
Comment (jpw547-zz) [reply fakebook/src/app/services/newsfeed.service.ts:42]: Fixed according to suggestions.

**Comment by jpw547-zz** [comment]
> don't request review when the build fails, let alone when the code doesn't even compile. that tells me you're not even running ng serve to check if what you typed is good.

I should have waited for the build to finish first before reviewing, but I actually requested the review before it had failed. Not a good answer, but I hadn't realized that it had failed.

And secondly, the build actually failed because of `ng lint`. We had run `ng serve` locally and everything built fine, but had not run `ng lint` so we had no idea that the build would fail when hitting the pipeline.

**Comment by escalonn** [comment]
when i was first requested it was an ng build error (you can see it in the history above ^); but you did fix that before i was done writing the review, leaving the lint error (which is less of a concern of course)

File: fakebook/src/app/component/like-view/like-view.component.ts [hunk 11]
```diff
@@ -7,23 +7,25 @@ import { LikeService } from 'src/app/services/like.service';
   styleUrls: ['./like-view.component.css'],
 })
 export class LikeViewComponent {
-  @Input() count: number | undefined;
-  @Input() postId: number;
-  @Input() liked: boolean;
+  @Input() count!: number;
+  @Input() postId!: number;
+  @Input() liked!: boolean;
```
Comment (escalonn) [inline fakebook/src/app/component/like-view/like-view.component.ts]: let's use a different approach here, per [style guide](https://angular.io/guide/styleguide#initialize-inputs)

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts [hunk 12]
```diff
@@ -13,43 +13,42 @@ import { PostService } from '../../services/post.service';
...
+      } else {
         this.getPosts();
       }
-    })
+    });
   }
 
   ngOnInit(): void {
     this.getUser();
-    //this.getPosts();
```
Comment (escalonn) [inline fakebook/src/app/component/newsfeed/newsfeed.component.ts]: do we not want to get posts on init? how else will the newsfeed show things? i might be missing something

File: fakebook/src/app/services/comment.service.ts [hunk 13]
```diff
@@ -3,21 +3,21 @@ import { Injectable } from '@angular/core';
...
 import { Comment } from 'src/app/model/comment';
 import { environment } from 'src/environments/environment';
-import { Post } from '../model/post';
 
 @Injectable({
   providedIn: 'root',
 })
 export class CommentService {
   url = `${environment.baseUrls.posts}/api/comments`;
+  acceptHeader = 'application/json';
```
Comment (escalonn) [inline fakebook/src/app/services/comment.service.ts]: let's make both of these properties private readonly

**Review by escalonn — CHANGES_REQUESTED** [review]

File: fakebook/src/app/component/newsfeed/newsfeed.component.ts (hunk 12 above)
Comment (Josh0224) [reply fakebook/src/app/component/newsfeed/newsfeed.component.ts]: the get post logic was in the constructor. I've changed it to be in the on init.

**Review by escalonn — CHANGES_REQUESTED** [review]
address at least the new issues:
https://sonarcloud.io/project/issues?createdAfter=2021-03-31&createdBefore=2021-04-14&id=2011-fakebook-project3_ng&pullRequest=77&resolved=false
the 'subscribe' ones are [a bug](https://github.com/microsoft/TypeScript/issues/43053), so mark them false positive.

**Review by escalonn — CHANGES_REQUESTED** [review]
again, address at least the new issues, filtered in the link in the previous review.

**Review by escalonn — APPROVED** [review]

**Review by escalonn — CHANGES_REQUESTED** [review]
not up to date with target branch

**Review by escalonn — APPROVED** [review]

**Pull request merged**